### PR TITLE
tls codec: variable-length vectors

### PR DESCRIPTION
**Pull request by franziskuskiefer**
This implements the variable-length integer encoding from [RFC 9000] (quic) used to serialize variable-length vectors in quic and in [MLS].
It's opinionated because it doesn't allow anything longer than `u16::MAX` when deserialising (same as the other vectors), and it requires minimum lengths like MLS (unlike quic).

cc @kkohbrok @raphaelrobert if one of you wants to have a look at this.

[rfc 9000]: https://www.rfc-editor.org/rfc/rfc9000.html#name-variable-length-integer-enc
[mls]: https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-variable-size-vector-header

### CODE REVIEW

**Comment by raphaelrobert** [comment]
Why specifically `u16::MAX`? Is that to ensure it will work on 16bit platforms? If not for the latter, I would recommend using `u32::MAX` instead.

File: tls_codec/src/quic_vec.rs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, Error> {
+        let (length, len_len) = read_variable_length(bytes)?;
+        if length == 0 {
+            // An empty vector.
+            return Ok(Vec::new());
+        }
+
+        let mut result = Vec::new();
```
Comment (kkohbrok) [inline tls_codec/src/quic_vec.rs:86]: Wouldn't it make sense to use something like `with_capacity((length-len_len)/element.tls_serialized_len())` here?

**Comment by franziskuskiefer** [comment]
> Why specifically `u16::MAX`? Is that to ensure it will work on 16bit platforms? If not for the latter, I would recommend using `u32::MAX` instead.

It's just what's used everywhere else already. As long as no one is actually running into the limit I don't see a reason to bump it.

**Comment by raphaelrobert** [comment]
> > Why specifically `u16::MAX`? Is that to ensure it will work on 16bit platforms? If not for the latter, I would recommend using `u32::MAX` instead.
> 
> It's just what's used everywhere else already. As long as no one is actually running into the limit I don't see a reason to bump it.

In the context of MLS this is limiting. It means the maximum size of encrypted payload is 64kb. When you use the ratchet tree extension for example (where the public tree gets encrypted in its entirety), this would severely limit the group size (to maybe less than 1000 members, depending on the size of KeyPackages). 
I wasn't aware that this limit also existed for `Vec32`, and I think it's equally problematic.

**Review by kkohbrok — COMMENTED** [review]
Looks good to me in general, although I have to admit, I didn't follow the byte operations in detail.

File: tls_codec/src/quic_vec.rs (hunk 1 above)
Comment (franziskuskiefer) [reply tls_codec/src/quic_vec.rs:86]: We could. It's a trade-off between allocating unnecessary memory and speeding up things. As long as no one measures performance here and identifies this as a bottleneck I'm conservative and don't allocate unnecessary memory.

Comment (kkohbrok) [reply tls_codec/src/quic_vec.rs:86]: I see. Do I understand correctly, though, that memory would only be unnecessarily allocated if the encoded length is inaccurate?

Comment (franziskuskiefer) [reply tls_codec/src/quic_vec.rs:86]: No, any broken element in the vector will cause the function to abort. The scenario I have in mind here is for example a vector that claims to be extremely long where the first element doesn't deserialize correctly. Then we'd allocate a ton of memory here and bail trying to read the first one.

**Pull request merged**